File: kotlin-node/karakum/src/jsMain/resources/plugins/DiffieHellmanGroupPlugin.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default {
    setup(context) {
        this.sourceDiffieHellmanGroupNode = null
    },

    traverse(node) {
        if (
            ts.isClassDeclaration(node)
            && node.name
            && node.name.text === "DiffieHellman"
        ) {
            this.sourceDiffieHellmanGroupNode = node
        }
    },

    render(node, context, next) {
        if (
            ts.isVariableDeclaration(node)
            && ts.isIdentifier(node.name)
            && node.name.text === "DiffieHellmanGroup"
        ) {
            return ""
        }

        if (
            ts.isTypeAliasDeclaration(node)
            && node.name.text === "DiffieHellmanGroup"
        ) {
            return ""
        }

        if (
            ts.isInterfaceDeclaration(node)
            && node.name.text === "DiffieHellmanGroupConstructor"
        ) {
            const inheritanceModifierService = context.lookupService(karakum.inheritanceModifierServiceKey.get())

            const inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)

            const ownMembers = node.members
                .filter(member => !(
                    ts.isPropertySignature(member)
                    && ts.isIdentifier(member.name)
                    && member.name.text === "prototype"
                ))
                .map(member => next(member))
                .join("\n")

            const members = this.sourceDiffieHellmanGroupNode.members
                .filter(member => !(
                    ts.isMethodDeclaration(member)
                    && ts.isIdentifier(member.name)
                    && (
                        member.name.text === "setPublicKey"
                        || member.name.text === "setPrivateKey"
                    )
                ))
                .map(member => next(member))
                .join("\n")

            return `
${karakum.ifPresent(inheritanceModifier, it => `${it} `)}external class DiffieHellmanGroup {
${ownMembers}\n${members}
}
            `
        }

        return null
    },

    generate(context) {
        return []
    }
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/DiffieHellmanGroupPlugin.kt
package node.karakum.plugins

import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.extension.plugins.InheritanceModifierService
import io.github.sgrishchenko.karakum.extension.plugins.inheritanceModifierServiceKey
import js.array.ReadonlyArray
import arrow.core.raise.impure
import arrow.core.raise.nullable
import typescript.*

class DiffieHellmanGroupPlugin : Plugin {
    private lateinit var sourceDiffieHellmanGroupNode: ClassDeclaration

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = impure {
        ensure(isClassDeclaration(node))
        ensure(node.name?.text == "DiffieHellman")

        sourceDiffieHellmanGroupNode = node
    }

    override fun render(node: Node, context: Context, next: Render<Node>) = nullable {
        nullable {
            ensure(isVariableDeclaration(node))

            val name = node.name
            ensure(isIdentifier(name))
            ensure(name.text == "DiffieHellmanGroup")

            ""
        } ?: nullable {
            ensure(isTypeAliasDeclaration(node))
            ensure(node.name.text == "DiffieHellmanGroup")

            ""
        } ?: nullable {
            ensure(isInterfaceDeclaration(node))
            ensure(node.name.text == "DiffieHellmanGroupConstructor")

            val inheritanceModifierService =
                ensureNotNull(context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey))

            val inheritanceModifier = inheritanceModifierService.resolveInheritanceModifier(node, context)

            val ownMembers = node.members.asArray()
                .mapNotNull { member ->
                    nullable {
                        ensure(
                            nullable {
                                ensure(isPropertySignature(member))

                                val propertyName = member.name
                                ensure(isIdentifier(propertyName))
                                ensure(propertyName.text == "prototype")
                            } == null
                        )

                        next(member)
                    }
                }
                .filter { it.isNotEmpty() }
                .joinToString("\n")

            val members = sourceDiffieHellmanGroupNode.members.asArray()
                .mapNotNull { member ->
                    nullable {
                        ensure(
                            nullable {
                                ensure(isMethodDeclaration(member))

                                val methodName = member.name
                                ensure(isIdentifier(methodName))
                                ensure(
                                    methodName.text == "setPublicKey"
                                            || methodName.text == "setPrivateKey"
                                )
                            } == null
                        )

                        next(member)
                    }
                }
                .filter { it.isNotEmpty() }
                .joinToString("\n")

            """
                ${ifPresent(inheritanceModifier) { "$it " }}external class DiffieHellmanGroup {
                $ownMembers
                $members
                }
            """.trimIndent()
        }
    }

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
